File: cli-python/src/code_archaeology/commands/dig.py
import os

import click
from rich.console import Console

from ..core.git_analyzer import GitAnalyzer
from ..core.timeline import build_result
from ..core.reporter import format_timeline
from ..annotator.inserter import insert_annotation

console = Console()


@click.command()
@click.argument("file")
@click.argument("line_range")
@click.option("--format", "fmt", type=click.Choice(["text", "json"]), default="text", help="Output format")
@click.option("--annotate", "do_annotate", is_flag=True, help="Prompt to insert a history annotation")
def dig(file: str, line_range: str, fmt: str, do_annotate: bool):
    """Analyze the git history of a code segment."""
    parts = line_range.split("-")
    start_line = int(parts[0])
    end_line = int(parts[1]) if len(parts) > 1 else start_line

    analyzer = GitAnalyzer(os.getcwd())

    if not analyzer.is_git_repo():
        console.print("[red]Not a git repository. Run this command inside a git project.[/]")
        raise SystemExit(1)

    commits = analyzer.get_line_history(file, start_line, end_line)

    if not commits:
        console.print("No git history found for this code segment.")
        return

    result = build_result(file, start_line, end_line, commits)

    if fmt == "json":
        import json
        from dataclasses import asdict
        console.print_json(json.dumps(asdict(result), default=str))
    else:
        format_timeline(result)

    if do_annotate:
        _prompt_annotation(os.path.abspath(file), result)


def _prompt_annotation(file_path: str, result):
    if not click.confirm("\nInsert a history annotation above this code?", default=True):
        return

    style = click.prompt(
        "Annotation style",
        type=click.Choice(["brief", "detailed"]),
        default="detailed",
    )

    success, message = insert_annotation(file_path, result, style)
    if success:
        console.print(f"\n[green]✅ {message}[/]")
    else:
        console.print(f"\n[red]❌ {message}[/]")

File: cli-node/src/commands/dig.ts
import { resolve } from 'node:path';
import inquirer from 'inquirer';
import { GitAnalyzer } from '../core/git-analyzer.js';
import { buildResult } from '../core/timeline.js';
import { formatTimeline, formatJson } from '../core/reporter.js';
import { insertAnnotation } from '../annotator/inserter.js';
import { AnnotationOptions } from '../types.js';

export interface DigOptions {
  format?: 'text' | 'json';
  annotate?: boolean;
}

export async function dig(file: string, lineRange: string, options: DigOptions): Promise<void> {
  const [startStr, endStr] = lineRange.split('-');
  const startLine = parseInt(startStr);
  const endLine = endStr ? parseInt(endStr) : startLine;

  if (isNaN(startLine) || isNaN(endLine)) {
    console.error('Invalid line range. Use format: 10-50 or single line: 10');
    process.exit(1);
  }

  const filePath = resolve(file);
  const analyzer = new GitAnalyzer(process.cwd());

  if (!(await analyzer.isGitRepo())) {
    console.error('Not a git repository. Run this command inside a git project.');
    process.exit(1);
  }

  const commits = await analyzer.getLineHistory(file, startLine, endLine);

  if (commits.length === 0) {
    console.log('No git history found for this code segment.');
    return;
  }

  const result = buildResult(file, startLine, endLine, commits);

  if (options.format === 'json') {
    console.log(formatJson(result));
  } else {
    console.log(formatTimeline(result));
  }

  if (options.annotate) {
    await promptAnnotation(filePath, result);
  }
}

async function promptAnnotation(filePath: string, result: import('../types.js').ArchaeologyResult): Promise<void> {
  const { shouldAnnotate } = await inquirer.prompt([{
    type: 'confirm',
    name: 'shouldAnnotate',
    message: 'Insert a history annotation above this code?',
    default: true,
  }]);

  if (!shouldAnnotate) return;

  const { style } = await inquirer.prompt([{
    type: 'list',
    name: 'style',
    message: 'Annotation style:',
    choices: [
      { name: 'Brief (summary only)', value: 'brief' },
      { name: 'Detailed (with key decisions)', value: 'detailed' },
    ],
  }]);

  const options: AnnotationOptions = {
    style,
    includeAuthors: true,
    includeIssues: true,
  };

  const { success, message } = insertAnnotation(filePath, result, options);
  if (success) {
    console.log(`\n✅ ${message}`);
  } else {
    console.error(`\n❌ ${message}`);
  }
}
